README.md
## Features
* None

## HOW TO
* Now just test in linux
  * sudo open vscode
    * sudo code . --user-data-dir ~/tmp
  * create a file
    * ctrl + n 
  * F2
    * Change COM and baudRate
  * F3
    * start or stop COM

## TO DO
* Support uart common function
  * change com id and buadrate
  * open and close
  * support multi uart
  * display hex
  * save log
  * send data
* save data
* trace manage
  * display or not display special line
  * color special line
* plot data
* send data 

CHANGELOG.md
# Change Log

All notable changes to the "timi-uart" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/util.js
"use strict";
const fs = require('fs');
const os = require('os');
const path = require('path');
const vscode = require('vscode');
const exec = require('child_process').exec;
const util = {
    /**
     * 获取当前所在工程根目录，有3种使用方法：<br>
     * getProjectPath(uri) uri 表示工程内某个文件的路径<br>
     * getProjectPath(document) document 表示当前被打开的文件document对象<br>
     * getProjectPath() 会自动从 activeTextEditor 拿document对象，如果没有拿到则报错
     * @param {*} document
     */
    getProjectPath(document) {
        if (!document) {
            document = vscode.window.activeTextEditor ? vscode.window.activeTextEditor.document : null;
        }
        if (!document) {
            this.showError('当前激活的编辑器不是文件或者没有文件被打开！');
            return '';
        }
        const currentFile = (document.uri ? document.uri : document).fsPath;
        let projectPath = null;
        let workspaceFolders = vscode.workspace.workspaceFolders.map(item => item.uri.path);
        // 由于存在Multi-root工作区，暂时没有特别好的判断方法，先这样粗暴判断
        // 如果发现只有一个根文件夹，读取其子文件夹作为 workspaceFolders
        if (workspaceFolders.length == 1 && workspaceFolders[0] === vscode.workspace.rootPath) {
            const rootPath = workspaceFolders[0];
            var files = fs.readdirSync(rootPath);
            workspaceFolders = files.filter(name => !/^\./g.test(name)).map(name => path.resolve(rootPath, name));
            // vscode.workspace.rootPath会不准确，且已过时
            // return vscode.workspace.rootPath + '/' + this._getProjectName(vscode, document);
        }
        workspaceFolders.forEach(folder => {
            if (currentFile.indexOf(folder) === 0) {
                projectPath = folder;
            }
        });
        if (!projectPath) {
            this.showError('获取工程根路径异常！');
            return '';
        }
        return projectPath;
    },
    /**
     * 获取当前工程名
     */
    getProjectName: function (projectPath) {
        return path.basename(projectPath);
    },
    getPluginPath() {
    },
    /**
     * 将一个单词首字母大写并返回
     * @param {*} word 某个字符串
     */
    upperFirstLetter: function (word) {
        return (word || '').replace(/^\w/, m => m.toUpperCase());
    },
    /**
     * 将一个单词首字母转小写并返回
     * @param {*} word 某个字符串
     */
    lowerFirstLeter: function (word) {
        return (word || '').replace(/^\w/, m => m.toLowerCase());
    },
    /**
     * 全局日志开关，发布时可以注释掉日志输出
     */
    log: function (...args) {
        console.log(...args);
    },
    /**
     * 全局日志开关，发布时可以注释掉日志输出
     */
    error: function (...args) {
        console.error(...args);
    },
    /**
     * 弹出错误信息
     */
    showError: function (info) {
        vscode.window.showErrorMessage(info);
    },
    /**
     * 弹出提示信息
     */
    showInfo: function (info) {
        vscode.window.showInformationMessage(info);
    },
    findStrInFolder: function (folderPath, str) {
    },
    /**
     * 从某个文件里面查找某个字符串，返回第一个匹配处的行与列，未找到返回第一行第一列
     * @param filePath 要查找的文件
     * @param reg 正则对象，最好不要带g，也可以是字符串
     */
    findStrInFile: function (filePath, reg) {
        const content = fs.readFileSync(filePath, 'utf-8');
        reg = typeof reg === 'string' ? new RegExp(reg, 'm') : reg;
        // 没找到直接返回
        if (content.search(reg) < 0)
            return { row: 0, col: 0 };
        const rows = content.split(os.EOL);
        // 分行查找只为了拿到行
        for (let i = 0; i < rows.length; i++) {
            let col = rows[i].search(reg);
            if (col >= 0) {
                return { row: i, col };
            }
        }
        return { row: 0, col: 0 };
    },
    /**
     * 获取某个字符串在文件里第一次出现位置的范围，
     */
    getStrRangeInFile: function (filePath, str) {
        var pos = this.findStrInFile(filePath, str);
        return new vscode.Range(new vscode.Position(pos.row, pos.col), new vscode.Position(pos.row, pos.col + str.length));
    },
    /**
     * 简单的检测版本大小
     */
    checkVersion: function (version1, version2) {
        version1 = parseInt(version1.replace(/\./g, ''));
        version2 = parseInt(version2.replace(/\./g, ''));
        return version1 > version2;
    },
    /**
     * 获取某个扩展文件绝对路径
     * @param context 上下文
     * @param relativePath 扩展中某个文件相对于根目录的路径，如 images/test.jpg
     */
    getExtensionFileAbsolutePath: function (context, relativePath) {
        return path.join(context.extensionPath, relativePath);
    },
    /**
     * 获取某个扩展文件相对于webview需要的一种特殊路径格式
     * 形如：vscode-resource:/Users/toonces/projects/vscode-cat-coding/media/cat.gif
     * @param context 上下文
     * @param relativePath 扩展中某个文件相对于根目录的路径，如 images/test.jpg
     */
    getExtensionFileVscodeResource: function (context, relativePath) {
        const diskPath = vscode.Uri.file(path.join(context.extensionPath, relativePath));
        return diskPath.with({ scheme: 'vscode-resource' }).toString();
    },
    /**
     * 在Finder中打开某个文件或者路径
     */
    openFileInFinder: function (filePath) {
        if (!fs.existsSync(filePath)) {
            this.showError('文件不存在：' + filePath);
        }
        // 如果是目录，直接打开就好
        if (fs.statSync(filePath).isDirectory()) {
            exec(`open ${filePath}`);
        }
        else {
            // 如果是文件，要分开处理
            const fileName = path.basename(filePath);
            filePath = path.dirname(filePath);
            // 这里有待完善，还不知道如何finder中如何选中文件
            exec(`open ${filePath}`);
        }
    },
    /**
     * 在vscode中打开某个文件
     * @param {*} path 文件绝对路径
     * @param {*} text 可选，如果不为空，则选中第一处匹配的对应文字
     */
    openFileInVscode: function (path, text) {
        let options = undefined;
        if (text) {
            const selection = this.getStrRangeInFile(path, text);
            options = { selection };
        }
        vscode.window.showTextDocument(vscode.Uri.file(path), options);
    },
    /**
     * 用JD-GUI打开jar包
     */
    openJarByJdGui: function (jarPath) {
        // 如何选中文件有待完善
        const jdGuiPath = vscode.workspace.getConfiguration().get('eggHelper.jdGuiPath');
        if (!jdGuiPath) {
            this.showError('JD-GUI路径不能为空！');
            return;
        }
        if (!fs.existsSync(jdGuiPath)) {
            this.showError('您还没有安装JD-GUI，请安装完后到vscode设置里面找到HSF助手并进行路径配置。');
            return;
        }
        if (!fs.existsSync(jarPath)) {
            this.showError('jar包未找到：' + jarPath);
            return;
        }
        exec(`open ${jarPath} -a ${jdGuiPath}`);
    },
    /**
     * 使用默认浏览器中打开某个URL
     */
    openUrlInBrowser: function (url) {
        exec(`open '${url}'`);
    },
    /**
     * 递归遍历清空某个资源的require缓存
     * @param {*} absolutePath
     */
    clearRequireCache(absolutePath) {
        const root = require.cache[absolutePath];
        if (!root)
            return;
        if (root.children) {
            // 如果有子依赖项，先清空依赖项的缓存
            root.children.forEach(item => {
                this.clearRequireCache(item.id);
            });
        }
        delete require.cache[absolutePath];
    },
    /**
     * 动态require，和普通require不同的是，加载之前会先尝试删除缓存
     * @param {*} modulePath
     */
    dynamicRequire(modulePath) {
        this.clearRequireCache(modulePath);
        return require(modulePath);
    },
    /**
     * 读取properties文件
     * @param {*} filePath
     */
    readProperties(filePath) {
        const content = fs.readFileSync(filePath, 'utf-8');
        let rows = content.split(os.EOL);
        rows = rows.filter(row => row && row.trim() && !/^#/.test(row));
        const result = {};
        rows.forEach(row => {
            let temp = row.split('=');
            result[temp[0].trim()] = temp[1].trim();
        });
        return result;
    },
    /**
     * 比较2个对象转JSON字符串后是否完全一样
     * 特别注意，由于JS遍历对象的无序性（部分浏览器是按照添加顺序遍历的）导致同样的对象，
     * 转成JSON串之后反而不一样，所以这里采用其它方式实现。
     * @param {*} obj1
     * @param {*} obj2
     */
    jsonEquals(obj1, obj2) {
        let s1 = this.formatToSpecialJSON(obj1, '', true);
        let s2 = this.formatToSpecialJSON(obj2, '', true);
        return s1 === s2;
    }
};
module.exports = util;
//# sourceMappingURL=util.js.map

src/uart.ts
import * as vscode from 'vscode';
import { stringify } from 'querystring';
const SerialPort = require('serialport');
// import * as SerialPort from 'serialport';

const Readline = require('@serialport/parser-readline');

var comID = '';

// test uart
var port: any;

const g_baudRate = '921600';
// // parser will lost log
// // const parser = new Readline();
// // port.pipe(parser);

async function listCom() {
	let serial_com = ['A', 'B'];

	let abc = await SerialPort.list(function (err: Error, ports: Array<{manufacturer: undefined, pnpId: undefined, comName: string}>) {
		console.log('******listCom Start******');
		ports.forEach(function(port) {
			// console.log(port);

			// Save devices info
			if (typeof(port.pnpId) !== 'undefined') {
				console.log(port.comName);
				console.log(port.pnpId);
				console.log(port.manufacturer);
				// vscode.window.showInformationMessage('COM: ' + port.comName);
				serial_com.push(port.comName);
				console.log('**********************');
			}
		});
		console.log('******listCom End******');
	});
	
	console.log(serial_com);
	
    return serial_com;
}

async function selectCom(item: Array<string>) {
	let i = 0;
	const com = await vscode.window.showQuickPick(item, {
		placeHolder: 'eins, zwei or drei',
		onDidSelectItem: item => vscode.window.showInformationMessage(`Focus ${++i}: ${item}`)
    });
    
    console.log(com);
    // vscode.window.showInformationMessage(`Got: ${result}`);
    
    return com;
}

async function setBaudRate() {
	const baudRate = await vscode.window.showInputBox({
		value: g_baudRate,
		valueSelection: [0, g_baudRate.length],
		placeHolder: 'Set baudRate for this COM',
		validateInput: text => {
			// vscode.window.showInformationMessage(`Validating: ${text}`);
			// return Number(baudRate) === '123' ? 'Not 123!' : null;
			let val = Number(text);
			console.log(val);

			if (isNaN(val)) {
				return 'Invalid input';
			}

			if (val < 1200) {
				return 'baudRate >= 1200';
			}

			if (val > 3000000) {
				return 'baudRate <= 3000000';
			}

			return null;
		}
	});

	console.log(baudRate);

	let val = Number(baudRate);

	// Check the val is valiable

	return val;
}


module.exports = function(context: vscode.ExtensionContext) {
	console.log('Congratulations, your extension "timi-uart" is now active!');

	context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart_open_or_close', async function() {
		// The code you place here will be executed every time your command is executed

		// Test display
		// vscode.window.showInformationMessage('TIMI: Test start...');


		console.log('Open or Stop COM...START');

		var change_cnt = 0;
		var disp_str = '';


		if (comID === '') {
			console.log('COM is NULL');

			return;
		}

		console.log('COM: ' + comID);

		

		console.log('0...');

		if (port.isOpen) {
			port.close();
			console.log('Close COM...');
			vscode.window.showInformationMessage('Open COM...');

			return;
		}

		console.log('Open COM...');

		await port.open(function (err: Error) {
			if (err) {
				return console.log('Error opening COM: ', err.message);
			} else {
				console.log('Open COM...OK');
				vscode.window.showInformationMessage('Open COM...');
			}
		});

		// // The open event is always emitted
		// port.on('open', function() {
		// 	console.log('Open COM...');
		// 	vscode.window.showInformationMessage('Open COM...');
		// });
	
		// port.on('readable', function () {
		// 	console.log('Data:', port.read());
		// });

		// parser.on('data', function (data) {
		port.on('data', function (data: Buffer) {
			console.log('Data:', data);
			// vscode.window.showInformationMessage(data);
			// let doc = editor.document;
			let editor = vscode.window.activeTextEditor;

			if (editor) {
				let doc = editor.document;

				var end_pos = doc.lineCount;
				var jump_range = new vscode.Range(end_pos, 0, end_pos, 0);

				disp_str += data;
				
				editor.edit(editBuilder => {
					editBuilder.replace(new vscode.Position(end_pos, 0), disp_str);
					disp_str = '';
					
				});

				editor.revealRange(jump_range, vscode.TextEditorRevealType.Default);
			}
		});

		console.log('Open or Stop COM...END');
	}));

	context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart_select', async function () {
		console.log('Select COM...');

		let serial_com = await listCom();
		let com = await selectCom(serial_com);
		let baudRate_val = NaN;

		if (com === undefined) {
			comID = '';
		} else {
			comID = com;
			baudRate_val = await setBaudRate();
		}

		console.log('Select COM: ' + comID);
		console.log('Set baudRate: ' + baudRate_val);	

		// Check input...
		if ((comID !== '') && !isNaN(baudRate_val)) {
			port = new SerialPort(comID, {baudRate: baudRate_val, autoOpen: false});
			console.log("Create COM...");
		} 

		console.log('Select COM...END');
	}));
};

out/welcome.js
"use strict";
const vscode = require('vscode');
const util = require('./util');
const fs = require('fs');
const path = require('path');
/**
 * 从某个HTML文件读取能被Webview加载的HTML内容
 * @param {*} context 上下文
 * @param {*} templatePath 相对于插件根目录的html文件相对路径
 */
function getWebViewContent(context, templatePath) {
    const resourcePath = util.getExtensionFileAbsolutePath(context, templatePath);
    const dirPath = path.dirname(resourcePath);
    let html = fs.readFileSync(resourcePath, 'utf-8');
    // vscode不支持直接加载本地资源，需要替换成其专有路径格式，这里只是简单的将样式和JS的路径替换
    html = html.replace(/(<link.+?href="|<script.+?src="|<img.+?src=")(.+?)"/g, (m, $1, $2) => {
        return $1 + vscode.Uri.file(path.resolve(dirPath, $2)).with({ scheme: 'vscode-resource' }).toString() + '"';
    });
    return html;
}
/**
 * 执行回调函数
 * @param {*} panel
 * @param {*} message
 * @param {*} resp
 */
function invokeCallback(panel, message, resp) {
    console.log('回调消息：', resp);
    // 错误码在400-600之间的，默认弹出错误提示
    if (typeof resp == 'object' && resp.code && resp.code >= 400 && resp.code < 600) {
        util.showError(resp.message || '发生未知错误！');
    }
    panel.webview.postMessage({ cmd: 'vscodeCallback', cbid: message.cbid, data: resp });
}
/**
 * 存放所有消息回调函数，根据 message.cmd 来决定调用哪个方法
 */
const messageHandler = {
    getConfig(global, message) {
        const result = vscode.workspace.getConfiguration().get(message.key);
        invokeCallback(global.panel, message, result);
    },
    setConfig(global, message) {
        // 写入配置文件，注意，默认写入工作区配置，而不是用户配置，最后一个true表示写入全局用户配置
        vscode.workspace.getConfiguration().update(message.key, message.value, true);
        util.showInfo('修改配置成功！');
    }
};
module.exports = function (context) {
    context.subscriptions.push(vscode.commands.registerCommand('extension.demo.showWelcome', function (uri) {
        const panel = vscode.window.createWebviewPanel('testWelcome', // viewType
        "自定义欢迎页", // 视图标题
        vscode.ViewColumn.One, // 显示在编辑器的哪个部位
        {
            enableScripts: true,
        });
        let global = { panel };
        panel.webview.html = getWebViewContent(context, 'src/view/custom-welcome.html');
        panel.webview.onDidReceiveMessage(message => {
            if (messageHandler[message.cmd]) {
                messageHandler[message.cmd](global, message);
            }
            else {
                util.showError(`未找到名为 ${message.cmd} 回调方法!`);
            }
        }, undefined, context.subscriptions);
    }));
    const key = 'vscodePluginDemo.showTip';
    // 如果设置里面开启了欢迎页显示，启动欢迎页
    if (vscode.workspace.getConfiguration().get(key)) {
        vscode.commands.executeCommand('extension.demo.showWelcome');
    }
};
//# sourceMappingURL=welcome.js.map

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// const SerialPort = require('serialport');
// const Readline = require('@serialport/parser-readline');
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
exports.activate = function (context) {
    console.log('恭喜，您的扩展“vscode-plugin-demo”已被激活！');
    // console.log(vscode);
    require('./helloword')(context);
    require('./uart')(context);
    // context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart', () => {
    //     vscode.window.showInformationMessage('Hello World！你好，小茗同学！');
    // }));
    // // Use the console to output diagnostic information (console.log) and errors (console.error)
    // // This line of code will only be executed once when your extension is activated
    // console.log('Congratulations, your extension "timi-uart" is now active!');
    // // The command has been defined in the package.json file
    // // Now provide the implementation of the command with registerCommand
    // // The commandId parameter must match the command field in package.json
    // let disposable = vscode.commands.registerCommand('extension.timi_uart', () => {
    // 	// The code you place here will be executed every time your command is executed
    // 	// Test display
    // 	vscode.window.showInformationMessage('TIMI: Test start...');
    // 	// vscode.window.showErrorMessage("ERROR info");
    // 	// vscode.window.showInformationMessage('Try to select？', 'Yes', 'No', 'Other').then(result => {
    // 	// 	if (result === 'Yes') {
    // 	// 		console.log("Yes button");
    // 	// 	} else if (result === 'No') {
    // 	// 		console.log("No button");
    // 	// 	}
    // 	// });
    // 	vscode.window.setStatusBarMessage("Test display status bar");
    // 	// Test display end
    // 	var change_cnt = 0;
    // 	var disp_str = '';
    // 	console.log('List serial start...');
    // 	SerialPort.list(function (err: Error, ports: Array<{manufacturer: undefined, pnpId: undefined, comName: undefined}>) {
    // 		ports.forEach(function(port) {
    // 			// console.log("port: Start...")
    // 			// console.log(port);
    // 			// Save devices info
    // 			if (typeof(port.pnpId) !== "undefined") {
    // 				console.log(port.comName);
    // 				console.log(port.pnpId);
    // 				console.log(port.manufacturer);
    // 				vscode.window.showInformationMessage('COM: ' + port.comName);
    // 			}
    // 		  	// console.log("port: End...")
    // 		});
    // 	});
    // 	// Display devices info
    // 	// Select device
    // 	console.log('List serial end...');
    // 	// test uart
    // 	const port = new SerialPort('/dev/ttyUSB0', {baudRate: 921600, autoOpen: false});
    // 	// parser will lost log
    // 	// const parser = new Readline();
    // 	// port.pipe(parser);
    // 	port.open(function (err: Error) {
    // 		console.log(typeof(err));
    // 		if (err) {
    // 		  return console.log('Error opening port: ', err.message);
    // 		}
    // 		// Because there's no callback to write, write errors will be emitted on the port:
    // 		port.write('main screen turn on');
    // 	});
    // 	// The open event is always emitted
    // 	port.on('open', function() {
    // 		console.log("Open COM...");
    // 		vscode.window.showInformationMessage('Open COM...');
    // 	});
    // 	// port.on('readable', function () {
    // 	// 	console.log('Data:', port.read());
    // 	// });
    // 	// parser.on('data', function (data) {
    // 	port.on('data', function (data: Buffer) {
    // 		console.log('Data:', data);
    // 		// vscode.window.showInformationMessage(data);
    // 		// let doc = editor.document;
    // 		let editor = vscode.window.activeTextEditor;
    // 		if (editor) {
    // 			let doc = editor.document;
    // 			var end_pos = doc.lineCount;
    // 			var jump_range = new vscode.Range(end_pos, 0, end_pos, 0);
    // 			disp_str += data;
    // 			editor.edit(editBuilder => {
    // 				editBuilder.replace(new vscode.Position(end_pos, 0), disp_str);
    // 				disp_str = '';
    // 			});
    // 			editor.revealRange(jump_range, vscode.TextEditorRevealType.Default);
    // 		}
    // 	});
    // 	console.log("END...");
    // });
    // context.subscriptions.push(disposable);
};
/**
 * 插件被释放时触发
 */
exports.deactivate = function () {
    console.log('您的扩展“vscode-plugin-demo”已被释放！');
};
//# sourceMappingURL=extension.js.map

out/helloword.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = require("vscode");
module.exports = function (context) {
    // 注册HelloWord命令
    context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart_test', () => {
        vscode.window.showInformationMessage('This is used to test API...');
        // vscode.window.showErrorMessage("ERROR info");
        // vscode.window.showInformationMessage('Try to select？', 'Yes', 'No', 'Other').then(result => {
        // 	if (result === 'Yes') {
        // 		console.log("Yes button");
        // 	} else if (result === 'No') {
        // 		console.log("No button");
        // 	}
        // });
        // vscode.window.setStatusBarMessage('Test display status bar');
    }));
};
//# sourceMappingURL=helloword.js.map

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
// const SerialPort = require('serialport');
// const Readline = require('@serialport/parser-readline');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
exports.activate =  function (context: vscode.ExtensionContext) {
    console.log('恭喜，您的扩展“vscode-plugin-demo”已被激活！');
    // console.log(vscode);
	require('./helloword')(context);
	require('./uart')(context);

	// context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart', () => {
    //     vscode.window.showInformationMessage('Hello World！你好，小茗同学！');
    // }));

	// // Use the console to output diagnostic information (console.log) and errors (console.error)
	// // This line of code will only be executed once when your extension is activated
	// console.log('Congratulations, your extension "timi-uart" is now active!');

	// // The command has been defined in the package.json file
	// // Now provide the implementation of the command with registerCommand
	// // The commandId parameter must match the command field in package.json
	// let disposable = vscode.commands.registerCommand('extension.timi_uart', () => {
	// 	// The code you place here will be executed every time your command is executed

	// 	// Test display
	// 	vscode.window.showInformationMessage('TIMI: Test start...');

	// 	// vscode.window.showErrorMessage("ERROR info");

	// 	// vscode.window.showInformationMessage('Try to select？', 'Yes', 'No', 'Other').then(result => {
	// 	// 	if (result === 'Yes') {
	// 	// 		console.log("Yes button");
	// 	// 	} else if (result === 'No') {
	// 	// 		console.log("No button");
	// 	// 	}
	// 	// });

	// 	vscode.window.setStatusBarMessage("Test display status bar");
		

	// 	// Test display end

	// 	var change_cnt = 0;
	// 	var disp_str = '';

	// 	console.log('List serial start...');

		

	// 	SerialPort.list(function (err: Error, ports: Array<{manufacturer: undefined, pnpId: undefined, comName: undefined}>) {
	// 		ports.forEach(function(port) {
	// 			// console.log("port: Start...")
	// 			// console.log(port);

	// 			// Save devices info
	// 			if (typeof(port.pnpId) !== "undefined") {
	// 				console.log(port.comName);
	// 				console.log(port.pnpId);
	// 				console.log(port.manufacturer);
	// 				vscode.window.showInformationMessage('COM: ' + port.comName);
	// 			}

	// 		  	// console.log("port: End...")
	// 		});
	// 	});

	// 	// Display devices info

	// 	// Select device

	// 	console.log('List serial end...');

	// 	// test uart
	// 	const port = new SerialPort('/dev/ttyUSB0', {baudRate: 921600, autoOpen: false});

	// 	// parser will lost log
	// 	// const parser = new Readline();
	// 	// port.pipe(parser);

	// 	port.open(function (err: Error) {
	// 		console.log(typeof(err));
	// 		if (err) {
	// 		  return console.log('Error opening port: ', err.message);
	// 		}
		  
	// 		// Because there's no callback to write, write errors will be emitted on the port:
	// 		port.write('main screen turn on');
	// 	});

	// 	// The open event is always emitted
	// 	port.on('open', function() {
	// 		console.log("Open COM...");
	// 		vscode.window.showInformationMessage('Open COM...');
	// 	});
	
	// 	// port.on('readable', function () {
	// 	// 	console.log('Data:', port.read());
	// 	// });

	// 	// parser.on('data', function (data) {
	// 	port.on('data', function (data: Buffer) {
	// 		console.log('Data:', data);
	// 		// vscode.window.showInformationMessage(data);
	// 		// let doc = editor.document;
	// 		let editor = vscode.window.activeTextEditor;

	// 		if (editor) {
	// 			let doc = editor.document;

	// 			var end_pos = doc.lineCount;
	// 			var jump_range = new vscode.Range(end_pos, 0, end_pos, 0);

	// 			disp_str += data;
				
	// 			editor.edit(editBuilder => {
	// 				editBuilder.replace(new vscode.Position(end_pos, 0), disp_str);
	// 				disp_str = '';
					
	// 			});

	// 			editor.revealRange(jump_range, vscode.TextEditorRevealType.Default);
	// 		}
	// 	});

	// 	console.log("END...");
	// });

	// context.subscriptions.push(disposable);
};

/**
 * 插件被释放时触发
 */
exports.deactivate = function() {
    console.log('您的扩展“vscode-plugin-demo”已被释放！');
};


out/uart.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = require("vscode");
const SerialPort = require('serialport');
// import * as SerialPort from 'serialport';
const Readline = require('@serialport/parser-readline');
var comID = '';
// test uart
var port;
const g_baudRate = '921600';
// // parser will lost log
// // const parser = new Readline();
// // port.pipe(parser);
function listCom() {
    return __awaiter(this, void 0, void 0, function* () {
        let serial_com = ['A', 'B'];
        let abc = yield SerialPort.list(function (err, ports) {
            console.log('******listCom Start******');
            ports.forEach(function (port) {
                // console.log(port);
                // Save devices info
                if (typeof (port.pnpId) !== 'undefined') {
                    console.log(port.comName);
                    console.log(port.pnpId);
                    console.log(port.manufacturer);
                    // vscode.window.showInformationMessage('COM: ' + port.comName);
                    serial_com.push(port.comName);
                    console.log('**********************');
                }
            });
            console.log('******listCom End******');
        });
        console.log(serial_com);
        return serial_com;
    });
}
function selectCom(item) {
    return __awaiter(this, void 0, void 0, function* () {
        let i = 0;
        const com = yield vscode.window.showQuickPick(item, {
            placeHolder: 'eins, zwei or drei',
            onDidSelectItem: item => vscode.window.showInformationMessage(`Focus ${++i}: ${item}`)
        });
        console.log(com);
        // vscode.window.showInformationMessage(`Got: ${result}`);
        return com;
    });
}
function setBaudRate() {
    return __awaiter(this, void 0, void 0, function* () {
        const baudRate = yield vscode.window.showInputBox({
            value: g_baudRate,
            valueSelection: [0, g_baudRate.length],
            placeHolder: 'Set baudRate for this COM',
            validateInput: text => {
                // vscode.window.showInformationMessage(`Validating: ${text}`);
                // return Number(baudRate) === '123' ? 'Not 123!' : null;
                let val = Number(text);
                console.log(val);
                if (isNaN(val)) {
                    return 'Invalid input';
                }
                if (val < 1200) {
                    return 'baudRate >= 1200';
                }
                if (val > 3000000) {
                    return 'baudRate <= 3000000';
                }
                return null;
            }
        });
        console.log(baudRate);
        let val = Number(baudRate);
        // Check the val is valiable
        return val;
    });
}
module.exports = function (context) {
    console.log('Congratulations, your extension "timi-uart" is now active!');
    context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart_open_or_close', function () {
        return __awaiter(this, void 0, void 0, function* () {
            // The code you place here will be executed every time your command is executed
            // Test display
            // vscode.window.showInformationMessage('TIMI: Test start...');
            console.log('Open or Stop COM...START');
            var change_cnt = 0;
            var disp_str = '';
            if (comID === '') {
                console.log('COM is NULL');
                return;
            }
            console.log('COM: ' + comID);
            console.log('0...');
            if (port.isOpen) {
                port.close();
                console.log('Close COM...');
                vscode.window.showInformationMessage('Open COM...');
                return;
            }
            console.log('Open COM...');
            yield port.open(function (err) {
                if (err) {
                    return console.log('Error opening COM: ', err.message);
                }
                else {
                    console.log('Open COM...OK');
                    vscode.window.showInformationMessage('Open COM...');
                }
            });
            // // The open event is always emitted
            // port.on('open', function() {
            // 	console.log('Open COM...');
            // 	vscode.window.showInformationMessage('Open COM...');
            // });
            // port.on('readable', function () {
            // 	console.log('Data:', port.read());
            // });
            // parser.on('data', function (data) {
            port.on('data', function (data) {
                console.log('Data:', data);
                // vscode.window.showInformationMessage(data);
                // let doc = editor.document;
                let editor = vscode.window.activeTextEditor;
                if (editor) {
                    let doc = editor.document;
                    var end_pos = doc.lineCount;
                    var jump_range = new vscode.Range(end_pos, 0, end_pos, 0);
                    disp_str += data;
                    editor.edit(editBuilder => {
                        editBuilder.replace(new vscode.Position(end_pos, 0), disp_str);
                        disp_str = '';
                    });
                    editor.revealRange(jump_range, vscode.TextEditorRevealType.Default);
                }
            });
            console.log('Open or Stop COM...END');
        });
    }));
    context.subscriptions.push(vscode.commands.registerCommand('extension.timi_uart_select', function () {
        return __awaiter(this, void 0, void 0, function* () {
            console.log('Select COM...');
            let serial_com = yield listCom();
            let com = yield selectCom(serial_com);
            let baudRate_val = NaN;
            if (com === undefined) {
                comID = '';
            }
            else {
                comID = com;
                baudRate_val = yield setBaudRate();
            }
            console.log('Select COM: ' + comID);
            console.log('Set baudRate: ' + baudRate_val);
            // Check input...
            if ((comID !== '') && !isNaN(baudRate_val)) {
                port = new SerialPort(comID, { baudRate: baudRate_val, autoOpen: false });
                console.log("Create COM...");
            }
            console.log('Select COM...END');
        });
    }));
};
//# sourceMappingURL=uart.js.map